core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -170,8 +171,10 @@ public synchronized void onDataReceived() {
       }
       // schedule a new ping
       state = State.PING_SCHEDULED;
-      pingFuture =
-          scheduler.schedule(sendPing, keepAliveTimeInNanos, TimeUnit.NANOSECONDS);
+      if (pingFuture == null) {
```

Comment:
Hm, that seems reasonable, but AFAIU `onDataReceived` can happen while we are in `State.IDLE/IDLE_WITH_PING_SENT` and so with your suggestion we would move to `PING_SCHEDULED` when really we might want to stay in `IDLE`. (Or something like that.)

I did make your suggested change and the tests still pass though...